### PR TITLE
Add support for hex float literals

### LINKED ISSUE #709: Proposal: Support C99 hex float literals to accurately capture float values
Originally: https://github.com/gpuweb/WSL/issues/351

Decimal notation for floating point numbers is fine for common cases, but confusingly lossy.   E.g. 1.5 vs. 1.3
Sometimes you really want to control exactly which floating point number is used, e.g. for mathematical constants like approximations to Pi, or for testing things like NaN-handling.

C99 hex float syntax is a worked-out absolutely unambiguous syntax for getting exactly the floating point numbers that you want.  WGSL should support this additional syntax.

### PR DESCRIPTION
**Pull request by alan-baker**
Fixes #709

* Adds literal tokenization to support hex floats

### CODE REVIEW

**Review by litherum — COMMENTED** [review]
Looks like something that’s just 0x7 would be parsed as a float literal and not an int literal?

**Comment by alan-baker** [comment]
This is the hex float part:
`0x([0-9a-fA-F]*.?[0-9a-fA-F]+ | [0-9a-fA-F]+.[0-9a-fA-F]*)(p|P)(+|-)?[0-9]+`

The exponent suffix is non-optional in this tokenization. The parentheses are matched such that the hex-digit-sequences are a combination that must be preceded by 0x and following by the exponent.

Simplifying it:
`0x(_hex_digit_sequence_)(p|P)(+|-)?[0-9]+`

File: wgsl/index.bs [hunk 1]
```diff
@@ -132,13 +132,13 @@ TODO: What indicates the end of a line?  (E.g. A line ends at the next linefeed
   <thead>
     <tr><th>Token<th>Definition
   </thead>
-  <tr><td>`FLOAT_LITERAL`<td>`(-?[0-9]*.[0-9]+ | -?[0-9]+.[0-9]*)((e|E)(+|-)?[0-9]+)?`
+  <tr><td>`FLOAT_LITERAL`<td>`-?(([0-9]*.[0-9]+ | [0-9]+.[0-9]*((e|E)(+|-)?[0-9]+)?) | 0x([0-9a-fA-F]*.?[0-9a-fA-F]+ | [0-9a-fA-F]+.[0-9a-fA-F]*)(p|P)(+|-)?[0-9]+)`
```
Comment (dneto0) [inline wgsl/index.bs:135]: This is really dense.
I suggest breaking this up into two,  DECIMAL_FLOAT_LITERAL  and HEX_FLOAT_LITERAL.

Then in the 3 other places where  FLOAT_LITERAL is used, replace it by uses of both of those.

Comment (dneto0) [reply wgsl/index.bs:135]: This section needs more exposition about how to actually convert these literals into actual values.
I could take that as a followup.

**Review by dneto0 — CHANGES_REQUESTED** [review]
Please break FLOAT_LITERAL into two rules.

**Comment by dneto0** [comment]
Discussed in the WGSL meeting of [2021-02-16](https://docs.google.com/document/d/1gDaPdTVrZLht9PY4G3VADkaK3Cw2-OJUFzzNBFE2BvM/edit#heading=h.hp3f2zbslxr9).

Consensus to approve the content of this on the understanding that it reuses C99 syntax.

File: wgsl/index.bs [hunk 2]
```diff
@@ -132,7 +132,8 @@ TODO: What indicates the end of a line?  (E.g. A line ends at the next linefeed
   <thead>
     <tr><th>Token<th>Definition
   </thead>
-  <tr><td>`FLOAT_LITERAL`<td>`-?(([0-9]*.[0-9]+ | [0-9]+.[0-9]*((e|E)(+|-)?[0-9]+)?) | 0x([0-9a-fA-F]*.?[0-9a-fA-F]+ | [0-9a-fA-F]+.[0-9a-fA-F]*)(p|P)(+|-)?[0-9]+)`
+  <tr><td>`DECIMAL_FLOAT_LITERAL`<td>`(-?[0-9]*.[0-9]+ | -?[0-9]+.[0-9]*)((e|E)(+|-)?[0-9]+)?`
+  <tr><td>`HEX_FLOAT_LITERAL`<td>`-?0x([0-9a-fA-F]*.?[0-9a-fA-F]+ | [0-9a-fA-F]+.[0-9a-fA-F]*)(p|P)(+|-)?[0-9]+`
```
Comment (dj2) [inline wgsl/index.bs:136]: Could we add a `<tr><td>`FLOAT_LITERAL`<td>DECIMAL_FLOAT_LITERAL | HEX_FLOAT_LITERAL and then leave the usages below?

File: wgsl/index.bs [hunk 3]
```diff
@@ -132,13 +132,14 @@ TODO: What indicates the end of a line?  (E.g. A line ends at the next linefeed
   <thead>
     <tr><th>Token<th>Definition
   </thead>
-  <tr><td>`FLOAT_LITERAL`<td>`(-?[0-9]*.[0-9]+ | -?[0-9]+.[0-9]*)((e|E)(+|-)?[0-9]+)?`
+  <tr><td>`DECIMAL_FLOAT_LITERAL`<td>`(-?[0-9]*.[0-9]+ | -?[0-9]+.[0-9]*)((e|E)(+|-)?[0-9]+)?`
+  <tr><td>`HEX_FLOAT_LITERAL`<td>`-?0x([0-9a-fA-F]*.?[0-9a-fA-F]+ | [0-9a-fA-F]+.[0-9a-fA-F]*)(p|P)(+|-)?[0-9]+`
```
Comment (dneto0) [inline wgsl/index.bs:136]: Comparing to C99 draft (http://www.open-std.org/jtc1/sc22/WG14/www/docs/n1256.pdf 6.4.42):

- should allow both 0x and 0X.  This worth raising to the group in a new issue.  I filed #1453
- the significand expression can be simplified to:
      
        ([0-9a-fA-F]*.[0-9a-fA-F]+ | [0-9a-fA-F]+.)

In fact, the same simplification can be applied to the decimal float literal.

- C99 doesn't include the leading minus.

The last two points are already tracked in https://github.com/gpuweb/gpuweb/issues/775

**Review by dneto0 — APPROVED** [review]
Further fixes are already tracked elsewhere.
Approving as this meets the goal of adding hex float literals.

**Pull request merged**